File: Libs/pricing/src/cpp/option.cpp

#include "option.hpp"

namespace QR 
{

Option::Option() :
_dayCounter(QuantLib::Actual365Fixed()), 
_initFlag(0) {} 

Option::Option(const Option& option_) 
{
    *this = option_; 
} 

Option& Option::operator=(const Option& option_) 
{
    if (this == &option_) 
        return *this;  
	_type         = option_._type; 
    _valueDate    = option_._valueDate; 
    _maturityDate = option_._maturityDate;  
    _S            = option_._S; 
    _K            = option_._K; 
    _T            = option_._T; 
    _sigma        = option_._sigma; 
    _r            = option_._r; 
    _q            = option_._q; 
    _dayCounter   = option_._dayCounter; 
    _initFlag     = option_._initFlag;    
	_price        = option_._price; 
	_greeks       = option_._greeks; 
    return *this; 
} 

Option::Option(
    Type type_, 
    const QuantLib::Date&  valueDate_,               
    const QuantLib::Date&  maturityDate_,               
    double  S_, 
    double  K_,
    double  sigma_, 
    double  r_,
	double  q_) :   
    _type(type_),  
    _valueDate(valueDate_),  
    _maturityDate(maturityDate_),  
    _S (S_),  
    _K(K_),  
    _sigma(sigma_),  
    _r(r_),
    _q(q_),
    _dayCounter(QuantLib::Actual365Fixed()), 
    _initFlag(0) 
{
	_T = _dayCounter.yearFraction(_valueDate, _maturityDate);
}  

Option::Option(
    Type    type_, 
    double  S_, 
    double  K_,
    double  T_, 
    double  sigma_, 
    double  r_,
	double  q_) :
    _type(type_),  
    _S(S_),  
    _K(K_),  
    _T(T_),  
    _sigma(sigma_),  
    _r(r_),      
    _q(q_),
    _dayCounter(QuantLib::Actual365Fixed()), 
    _initFlag(0) {}
    
std::string Option::getTypeName() const 
{
    if (_type == PUT) 
        return std::string("PUT");
    else
        return std::string("CALL");  
}

void Option::calcDelta(double pct_) 
{
	OptionPtr option1(this->clone());
	OptionPtr option2(this->clone());
	option1->setSpot( (1 - pct_) * _S );
	option2->setSpot( (1 + pct_) * _S );
	option1->calcPrice(); 
	option2->calcPrice();
	double delta = ( option2->getPrice() - option1->getPrice() ) / ( 2 * pct_ * _S ); 
	_greeks.setDelta(delta); 
}

void Option::calcVega(double pct_) 
{
	OptionPtr option1(this->clone());
	OptionPtr option2(this->clone());
	option1->setVol( (1 - pct_) * _sigma );
	option2->setVol( (1 + pct_) * _sigma );
	option1->calcPrice(); 
	option2->calcPrice();
	double vega = ( option2->getPrice() - option1->getPrice() ) / ( 2 * pct_ * _sigma ); 
	_greeks.setVega(vega); 
}

void Option::calcGamma(double pct_) 
{
	OptionPtr option1(this->clone());
	OptionPtr option2(this->clone());
	option1->setSpot( (1 - pct_) * _S );
	option2->setSpot( (1 + pct_) * _S );
	option1->calcPrice(); 
	option2->calcPrice();
	calcPrice(); 
	double gamma = ( option1->getPrice() + option2->getPrice() - 2 * _price ) / ( pct_ * _S * pct_ * _S ); 
	_greeks.setGamma(gamma); 
}

void Option::calcTheta(double pct_)
{
	OptionPtr option1(this->clone());
	OptionPtr option2(this->clone());
	option1->setT2M( (1 - pct_) * _T );
	option2->setT2M( (1 + pct_) * _T );
	option1->calcPrice(); 
	option2->calcPrice();
	double theta = ( option1->getPrice() - option2->getPrice() ) / ( 2 * pct_ * _T ); 
	_greeks.setTheta(theta); 
}

void Option::calcRho(double pct_)
{
	OptionPtr option1(this->clone());
	OptionPtr option2(this->clone());
	option1->setRate( (1 - pct_) * _r );
	option2->setRate( (1 + pct_) * _r );
	option1->calcPrice(); 
	option2->calcPrice();
	double rho = ( option2->getPrice() - option1->getPrice() ) / ( 2 * pct_ * _r ); 
	_greeks.setRho(rho); 
}

double Option::calcImplVol(double price_) 
{
	QuantLib::Bisection bisection; 
	double accuracy = 1.e-6;
	double guess = 0.15; 
	double xmin = 1.e-5; 
	double xmax = 5.; 
	double impl_vol = bisection.solve(
					FunctorImplVol(this, price_), 
					accuracy, 
					guess, 
					xmin, 
					xmax); 
	return impl_vol; 
}

std::ostream& operator<<(std::ostream& out_, const Option& option_) 
{
    out_ << "Type=" << option_.getTypeName() << std::endl; 
    out_ << "ValueDate=" << option_.getValueDate() << std::endl; 
    out_ << "MaturityDate=" << option_.getMaturityDate() << std::endl; 
    out_ << "Spot=" << option_.getSpot() << std::endl; 
    out_ << "Strike=" << option_.getStrike() << std::endl; 
    out_ << "T2M=" << option_.getT2M() << std::endl; 
    out_ << "Volatility=" << option_.getVol() << std::endl; 
    out_ << "Risk free rate=" << option_.getRate() << std::endl; 
    out_ << "Dividend yield=" << option_.getDivYield() << std::endl; 
    out_ << "DayCounter=" << option_.getDCC().name() << std::endl; 
    out_ << "InitFlag=" << option_.getInitFlag() << std::endl; 
    out_ << "Price=" << option_.getPrice() << std::endl; 
    out_ << "Greeks=" << option_.getGreeks() << std::endl; 
    return out_; 
}

double FunctorImplVol::operator() (double vol_) const 
{
	OptionPtr optPtr(_optionPtr->clone());
	optPtr->setVol(vol_);
	optPtr->calcPrice();
	double f = optPtr->getPrice() - _price;
	return f;
}

}

File: Libs/pricing/src/py/option.py
#!/usr/bin/python
###########################################################
#  python option class 
#  convert python option class to C++ option class 
###########################################################

import os
import liboqr_py as oqr
import option_util as util

class AmOption(oqr.AmOption):
	PCT_CHG_GREEKS = .001

	def __init__(self, cp_type_ = "", trade_date_ = 0, exp_date_ = 0, \
			T_ = 0., S_ = 0., K_ = 0., sigma_ = 0., r_ =0., q_ = 0., price_impl_vol_ = 0.): 
		if trade_date_ != 0 and exp_date_ != 0:
			oqr.AmOption.__init__(self, util.str2cp_type(cp_type_), util.str2qlDate(str(trade_date_)), \
							util.str2qlDate(str(exp_date_)), S_, K_, sigma_, r_, q_) 
		else:
			oqr.AmOption.__init__(self, util.str2cp_type(cp_type_), S_, K_, T_, sigma_, r_, q_) 

		self.price_impl_vol = price_impl_vol_
		self.impl_vol       = .0
		self.delta          = .0
		self.vega           = .0
		self.gamma          = .0
		self.theta          = .0
		self.rho            = .0 

	def setSigma(self, sigma_):
		self.sigma = sigma_

	def print_out(self):
		print "ex_style=AMERICAN", "cp_type=", util.cp_type2str(self.cp_type), \
			"trade_date=", util.qlDate2str(self.valueDate), "exp_date=", util.qlDate2str(self.maturityDate), \
			"T=", self.T, "S=", self.S, "K=", self.K, "sigma=", self.sigma, "r=", self.r, "q=", self.q, \
			"price=", self.price, "price_impl_vol=", self.price_impl_vol, "impl_vol=", self.impl_vol 
	
	def print_greeks(self): 
		print "delta=%f gamma=%f vega=%f theta=%f rho=%f" %(self.delta, self.gamma, self.vega, self.theta, self.rho) 

	def calcPrice(self): 
		super(AmOption, self).calcPrice()
		return self.price 	

	def calcImplVol(self, price_ = 0.):
		if price_ <= 0.: 
			self.impl_vol = super(AmOption, self).calcImplVol(self.price_impl_vol)
		else:
			self.impl_vol = super(AmOption, self).calcImplVol(price_)
		return self.impl_vol 

	def calcDelta(self): 
		oqr.Option.calcDelta(self, self.PCT_CHG_GREEKS)
		self.delta = self.greeks.delta
		return self.delta

	def calcVega(self): 
		oqr.Option.calcVega(self, self.PCT_CHG_GREEKS)
		self.vega = self.greeks.vega
		return self.vega
	
	def calcGamma(self): 
		oqr.Option.calcGamma(self, self.PCT_CHG_GREEKS)
		self.gamma = self.greeks.gamma
		return self.gamma

	def calcTheta(self): 
		oqr.Option.calcTheta(self, self.PCT_CHG_GREEKS)
		self.theta = self.greeks.theta
		return self.theta

	def calcRho(self): 
		oqr.Option.calcRho(self, self.PCT_CHG_GREEKS)
		self.rho = self.greeks.rho
		return self.rho

if __name__ == "__main__": 
#	option1 = Option(ex_style_ = "AMERICAN", cp_type_ = "PUT", \
#		T_ = 1., S_ = 36, K_ = 40, sigma_ = .2, r_ = .06, q_ = .06, price_impl_vol_ = 6) 
#	option1.calcPrice()	
#	option1.calcImplVol() 
#	option1.print_out()
#	print "price=", option1.price, "impl_vol=", option1.impl_vol 

	option2 = AmOption(cp_type_ = "CALL", trade_date_ = 20140301, exp_date_ = 20140321, \
			S_ = 1866., K_ = 1865., sigma_ = .0, r_ = .00155, q_ = .00155, price_impl_vol_ = 11.121) 
	option2.print_out()
	iv = option2.calcImplVol() 
	option2.setSigma(iv)
	option2.calcDelta() 
	option2.calcVega() 
	option2.calcGamma() 
	option2.calcTheta() 
	option2.calcRho() 
	print "#############IV=", iv
	option2.calcPrice()
	option2.print_out()
	option2.print_greeks()

